bottlecap/src/bin/bottlecap/main.rs

```diff
@@ -486,17 +523,163 @@ async fn extension_loop_active(
-                stats_flusher.flush()
-            );
+            flush_all(
+                &logs_flusher,
+                &mut metrics_flusher,
+                &*trace_flusher,
+                &*stats_flusher,
+                &mut flush_interval,
+            )
+            .await;
             return Ok(());
         }
     }
 }
 
+async fn flush_all(
+    logs_flusher: &LogsFlusher,
+    metrics_flusher: &mut MetricsFlusher,
+    trace_flusher: &dyn TraceFlusher,
+    stats_flusher: &dyn StatsFlusher,
+    flush_interval: &mut tokio::time::Interval,
+) {
+    tokio::join!(
+        logs_flusher.flush(),
+        metrics_flusher.flush(),
+        trace_flusher.flush(),
+        stats_flusher.flush()
+    );
+    flush_interval.reset();
+}
+
+struct RuntimeDoneMeta {
+    request_id: String,
+    status: Status,
+    metrics: RuntimeDoneMetrics,
+}
+
+async fn handle_event_bus_event(
+    event: Event,
+    invocation_processor: Arc<TokioMutex<InvocationProcessor>>,
+) -> Option<RuntimeDoneMeta> {
+    match event {
+        Event::Metric(event) => {
+            debug!("Metric event: {:?}", event);
+            None
+        }
+        Event::OutOfMemory => {
+            let mut p = invocation_processor.lock().await;
+            p.on_out_of_memory_error();
+            drop(p);
+            None
+        }
+        Event::Telemetry(event) => match event.record {
+            TelemetryRecord::PlatformInitStart { .. } => {
+                let mut p = invocation_processor.lock().await;
+                p.on_platform_init_start(event.time);
+                drop(p);
+                None
+            }
+            TelemetryRecord::PlatformInitReport {
+                initialization_type,
+                phase,
+                metrics,
+            } => {
+                debug!("Platform init report for initialization_type: {:?} with phase: {:?} and metrics: {:?}", initialization_type, phase, metrics);
+                let mut p = invocation_processor.lock().await;
+                p.on_platform_init_report(metrics.duration_ms);
+                drop(p);
+                None
+            }
+            TelemetryRecord::PlatformStart { request_id, .. } => {
+                let mut p = invocation_processor.lock().await;
+                p.on_platform_start(request_id, event.time);
+                drop(p);
+                None
+            }
+            TelemetryRecord::PlatformRuntimeDone {
+                request_id,
+                status,
+                metrics,
+                error_type,
+                ..
+            } => {
+                debug!(
+                    "Runtime done for request_id: {:?} with status: {:?} and error: {:?}",
+                    request_id,
+                    status,
+                    error_type.unwrap_or("None".to_string())
+                );
+
+                if let Some(metrics) = metrics {
+                    return Some(RuntimeDoneMeta {
```

Comment:
Can't we just sent the event back?

Reply:
yeah that's probably better

Reply:
I don't like it a 100% but we can change it in the future. My idea was still sending the clone of the event as an optional, that way we ensure we don't have to push the conditional back to the loop.

It just feels unnecessary when an object already exist for this which has the same information.

Reply:
yeah it's a little weirder with the enum because we strip a bunch of the event info away and just return what we need. It's another alloc but it should be stack alloc'd and fast anyway. We can change it later